testng-core/src/main/java/org/testng/internal/TestNGClassFinder.java

```diff
@@ -152,7 +149,8 @@ private void processMethod(
 
   private static boolean excludeFactory(FactoryMethod fm, ITestContext ctx) {
     return fm.getGroups().length != 0
-        && ctx.getCurrentXmlTest().getExcludedGroups().containsAll(Arrays.asList(fm.getGroups()));
+        && new HashSet<>(ctx.getCurrentXmlTest().getExcludedGroups())
+            .containsAll(Arrays.asList(fm.getGroups()));
```

Comment:
Maybe use hashset instead of list here too

Reply:
I read through on this and it looks like the changeset that I have done is sufficient because the `HashSet<>` implementation short circuits the `contains()` implementation and thus reduces the nested iteration.

SO eference: https://stackoverflow.com/a/39575190